### PR TITLE
[DS-362] - Create banner component  🚀

### PR DESCRIPTION
**Pull request by ericcleao**
[JIRA Issue](https://gympass.atlassian.net/browse/DS-362)

## Description 📄

- Creation of the Banner component as specified at the Banner Figma branch on the Yoga;

## Platforms 📲

- [x] Web
- [x] Mobile

## Type of change 🔍

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested? 🧪

- Check the documentation what the component looks like;

- [x] Unit Test
- [x] Snapshot Test

## Checklist: 🔍

- [x] My code follows the contribution guide of this project [Contributing Guide](https://github.com/Gympass/yoga/blob/master/CONTRIBUTING.md)
- [x] Layout matches design prototype: [FIGMA](https://www.figma.com/file/wn9grDi9mM29Wt7LmQjol7/branch/MN3NDJBxBKZrS00LOAF82p/%E2%9C%85-Yoga-%5BMAIN%5D?node-id=24374%3A254240)
- [x] I have performed a self-review of my code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### Web Doc

image

image

image

image

image

### LabNative

| iOS | Android |
| ------ | ----- |
| Screen Recording 2022-04-29 at 01 06 31 | Screen Recording 2022-04-29 at 00 32 22 |

### CODE REVIEW

**Review by fjpamaral — APPROVED** [review]

**Comment by Wendlereis** [comment]
How the consumer application should control its visibility? Should we have a `open` prop?

**Comment by ericcleao** [comment]
> With that in mind, is it really design-ready and waiting for development?

Seems that the `anatomy`, `placement`, and `behavior` are still in `wip` at the design side. It was aligned with the design team about coding the component in parallel and delivering it together, I'm going to ask the design team about those three missing sections and will bring the answer back here.

**Comment by ericcleao** [comment]
> How the consumer application should control its visibility? Should we have a open prop?

The `Banner`, at least for now, is not closable.
It should be used on the design side to show to the user a message the is about some status and this message should be there while the status remains, if the status changes the component should not be rendered, like:

```jsx
{!!isDowngradeScheduled && (
  <Banner
    message={scheduledDowngrade.label}
    button={{
      label: scheduledDowngrade.actionLabel,
      action: () => {
        tracker.trackEvent('home_undo_downgrade_clicked', {
          category: 'home',
        });
        navigation.navigate('Membership', { category: 'home' });
      },
    }}
  />
)}
```

**Comment by Wendlereis** [comment]
> The `Banner`, at least for now, is not closable.

What do you think to go more declarative and let the component handle the visibility internally?

```
  <Banner
    open={!!isDowngradeScheduled}
    message={scheduledDowngrade.label}
    button={{
      label: scheduledDowngrade.actionLabel,
      action: () => {
        tracker.trackEvent('home_undo_downgrade_clicked', {
          category: 'home',
        });
        navigation.navigate('Membership', { category: 'home' });
      },
    }}
  />
```

**Comment by ericcleao** [comment]
> What do you think to go more declarative and let the component handle the visibility internally?

I like it because it put inside the component the visibility handle, but it should not be a common behavior, it suggests to the developer that the component could be opened at a time and be closed after a button click, for example:

```jsx
 <Banner
    open={!!bannerIsOpen} // <-- the prop
    message="Banner message"
    button={{
      label: "close",
      action: () => {
        setBannerIsOpen(false);
      },
    }}
  />
```

and it should not be a usage of the component, for that usage, it will better the Snackbar.

**Comment by ericcleao** [comment]
> What do you think to go more declarative and let the component handle the visibility internally?

Not rendering the Banner component based on some `status` is a business rule, but if we add a prop to the component itself that allows `closing` it, we make it able to some behavior that should be avoided.

**Comment by mmartinsoliv** [comment]
> 

I believe it is not necessary to have an `open` prop, the banner behavior is expected to be "fixed"

**Comment by Wendlereis** [comment]
> ```js
> button={{
>       label: "close",
>       action: () => {
>         setBannerIsOpen(false);
>       },
>     }}
> ```



> > What do you think to go more declarative and let the component handle the visibility internally?
> 
> Not rendering the Banner component based on some `status` is a business rule, but if we add a prop to the component itself that allows `closing` it, we make it able to some behavior that should be avoided at the design side.

Got it, makes sense. Thanks

**Review by mmartinsoliv — APPROVED** [review]

File: packages/yoga/src/Banner/web/Banner.test.jsx [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+      <ThemeProvider>
+        <Banner
+          message="Banner with button"
+          actionLabel="Action"
+          onAction={onButtonClickMock}
+        />
+      </ThemeProvider>,
+    );
+
+    fireEvent.click(getByText('Action'));
```
Comment (frgiovanna) [inline packages/yoga/src/Banner/web/Banner.test.jsx:58]:
```suggestion
    const { getByRole } = render(
      <ThemeProvider>
        <Banner
          message="Banner with button"
          actionLabel="Action"
          onAction={onButtonClickMock}
        />
      </ThemeProvider>,
    );

    fireEvent.click(getByRole('button', { name: 'Action' }));
```

**Review by frgiovanna — APPROVED** [review]

File: packages/yoga/src/Banner/web/Banner.test.jsx (hunk 1 above)
Comment (ericcleao) [reply packages/yoga/src/Banner/web/Banner.test.jsx]: Done! At the native I could not follow the same approach, maybe we could add `accessibilityRole` to the native `Button.Text` to improve the tests and the accessibility as a whole, @mmartinsoliv.

**Review by mmartinsoliv — APPROVED** [review]

**Comment by ericcleao** [comment]
Implementing changes that were inserted at the component while design discovery.

**Review by lguilhermecardoso — APPROVED** [review]

File: packages/yoga/src/Banner/native/Banner.jsx [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        <Icon as={icon} size="medium" fill="secondary" marginRight="xxsmall" />
+      )}
+      <Text.Small flex={1} marginVertical="xxsmall">
+        {message}
+      </Text.Small>
+      {!!primaryButton && !secondaryButton && (
+        <BannerActionButton button={primaryButton} marginLeft="xxsmall" />
+      )}
+    </Box>
+    {primaryButton && secondaryButton && (
```
Comment (renanmpn) [inline packages/yoga/src/Banner/native/Banner.jsx:84]: why the first one (80) it's `!!primaryButton` and the second (84) `primaryButton`?

**Review by renanmpn — APPROVED** [review]

File: packages/yoga/src/Banner/native/Banner.jsx (hunk 2 above)
Comment (ericcleao) [reply packages/yoga/src/Banner/native/Banner.jsx:84]: For any specific reason, I'm going to update it to use the same pattern at both comparison

Comment (ericcleao) [reply packages/yoga/src/Banner/native/Banner.jsx]: Done!

File: packages/yoga/src/Banner/native/Banner.jsx [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    return checkPropTypes(
+      { [propName]: BannerActionButtonType },
+      props,
+      'prop',
+      componentName,
+    );
+  },
+};
```
[the remark this answers is not in the document]
Comment (ericcleao) [reply packages/yoga/src/Banner/native/Banner.jsx:139]: Thanks!! Thinking about if it is interesting to put this propType to `packages/yoga/src/shared/propTypes`, to be used in other contexts or just do that in the future if it is shown to be necessary.

File: packages/yoga/src/Banner/web/Banner.jsx [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+const Banner = ({
+  icon,
+  variant,
+  message,
+  primaryButton,
+  secondaryButton,
+  ...props
+}) => (
+  <StyledBanner variant={variant} {...props}>
+    <Box display="flex" flex={1} flexDirection="row" alignItems="center">
```
[the remark this answers is not in the document]
Comment (ericcleao) [reply packages/yoga/src/Banner/web/Banner.jsx]: Removed!

File: packages/yoga/src/Banner/native/Banner.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (mmartinsoliv) [reply packages/yoga/src/Banner/native/Banner.jsx:139]: It's a great idea.

**Pull request merged**